File: apps/api/app/routers/webhooks.py
import hashlib
import hmac
import secrets
from typing import Any

import httpx
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from app.auth import get_current_user
from app.deps import optional_supabase

router = APIRouter(prefix="/webhooks", tags=["webhooks"])


class WebhookBody(BaseModel):
    url: str
    events: list[str]


def _sign(payload: str, secret: str) -> str:
    return hmac.new(secret.encode(), payload.encode(), hashlib.sha256).hexdigest()


async def _tenant_id(user: dict, db: Any) -> str:
    row = (
        db.table("nano_tenants")
        .select("id")
        .eq("owner_phone", user["phone_number"])
        .maybe_single()
        .execute()
    )
    if not row.data:
        raise HTTPException(404, "Tenant not found")
    return row.data["id"]


@router.get("")
async def list_webhooks(
    user: dict = Depends(get_current_user),
    db: Any = Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    tid = await _tenant_id(user, db)
    row = (
        db.table("nano_webhooks")
        .select("id,url,events,is_active,last_fired_at,created_at")
        .eq("tenant_id", tid)
        .execute()
    )
    return row.data


@router.post("")
async def create_webhook(
    body: WebhookBody,
    user: dict = Depends(get_current_user),
    db: Any = Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    tid = await _tenant_id(user, db)
    secret = secrets.token_urlsafe(32)
    row = db.table("nano_webhooks").insert({
        "tenant_id": tid,
        "url": body.url,
        "events": body.events,
        "secret": secret,
    }).execute()
    data = row.data[0] if row.data else {}
    return {**data, "secret": secret}  # secret returned once on creation


@router.delete("/{webhook_id}")
async def delete_webhook(
    webhook_id: str,
    user: dict = Depends(get_current_user),
    db: Any = Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    tid = await _tenant_id(user, db)
    row = (
        db.table("nano_webhooks")
        .delete()
        .eq("id", webhook_id)
        .eq("tenant_id", tid)
        .execute()
    )
    if not row.data:
        raise HTTPException(404, "Webhook not found")
    return {"deleted": True}


@router.post("/{webhook_id}/test")
async def test_webhook(
    webhook_id: str,
    user: dict = Depends(get_current_user),
    db: Any = Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    tid = await _tenant_id(user, db)
    row = (
        db.table("nano_webhooks")
        .select("*")
        .eq("id", webhook_id)
        .eq("tenant_id", tid)
        .maybe_single()
        .execute()
    )
    if not row.data:
        raise HTTPException(404, "Webhook not found")
    wh = row.data
    payload = '{"event":"test","data":{"message":"Test webhook from Nanowork"}}'
    sig = _sign(payload, wh["secret"])
    async with httpx.AsyncClient(timeout=10.0) as client:
        try:
            resp = await client.post(
                wh["url"],
                content=payload,
                headers={
                    "Content-Type": "application/json",
                    "X-Nanowork-Signature": sig,
                },
            )
            db.table("nano_webhooks").update({"last_fired_at": "now()"}).eq("id", webhook_id).execute()
            return {"delivered": resp.status_code < 400, "status_code": resp.status_code}
        except Exception as exc:
            return {"delivered": False, "error": str(exc)}

File: apps/worker/src/routes/webhooks.ts
import { Hono } from 'hono'
import type { Env } from '../index'
import { getUser } from '../middleware/auth'
import { getSupabase } from '../lib/supabase'

const app = new Hono<{ Bindings: Env }>()

async function tenantId(user: Record<string, unknown>, env: Env): Promise<string | null> {
  const sb = getSupabase(env)
  const { data } = await sb.from('nano_tenants').select('id').eq('owner_phone', user.phone_number).maybeSingle()
  return data?.id ?? null
}

async function hmacHex(secret: string, payload: string): Promise<string> {
  const key = await crypto.subtle.importKey(
    'raw', new TextEncoder().encode(secret),
    { name: 'HMAC', hash: 'SHA-256' }, false, ['sign'],
  )
  const sig = await crypto.subtle.sign('HMAC', key, new TextEncoder().encode(payload))
  return Array.from(new Uint8Array(sig)).map(b => b.toString(16).padStart(2, '0')).join('')
}

function randomSecret(): string {
  const bytes = new Uint8Array(32)
  crypto.getRandomValues(bytes)
  return btoa(String.fromCharCode(...bytes)).replace(/[+/=]/g, c => ({ '+': '-', '/': '_', '=': '' }[c] ?? c))
}

app.get('/', async (c) => {
  let user: Record<string, unknown>
  try { user = await getUser(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const tid = await tenantId(user, c.env)
  if (!tid) return c.json({ error: 'Tenant not found' }, 404)
  const sb = getSupabase(c.env)
  const { data } = await sb.from('nano_webhooks').select('id,url,events,is_active,last_fired_at,created_at').eq('tenant_id', tid)
  return c.json(data ?? [])
})

app.post('/', async (c) => {
  let user: Record<string, unknown>
  try { user = await getUser(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const tid = await tenantId(user, c.env)
  if (!tid) return c.json({ error: 'Tenant not found' }, 404)
  const { url, events } = await c.req.json<{ url: string; events: string[] }>()
  const secret = randomSecret()
  const sb = getSupabase(c.env)
  const { data, error } = await sb.from('nano_webhooks').insert({ tenant_id: tid, url, events, secret }).select().single()
  if (error) return c.json({ error: error.message }, 503)
  return c.json({ ...data, secret })
})

app.delete('/:id', async (c) => {
  let user: Record<string, unknown>
  try { user = await getUser(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const tid = await tenantId(user, c.env)
  if (!tid) return c.json({ error: 'Tenant not found' }, 404)
  const sb = getSupabase(c.env)
  const { data, error } = await sb.from('nano_webhooks').delete().eq('id', c.req.param('id')).eq('tenant_id', tid).select()
  if (error) return c.json({ error: error.message }, 503)
  if (!data?.length) return c.json({ error: 'Webhook not found' }, 404)
  return c.json({ deleted: true })
})

app.post('/:id/test', async (c) => {
  let user: Record<string, unknown>
  try { user = await getUser(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const tid = await tenantId(user, c.env)
  if (!tid) return c.json({ error: 'Tenant not found' }, 404)
  const sb = getSupabase(c.env)
  const { data: wh } = await sb.from('nano_webhooks').select('*').eq('id', c.req.param('id')).eq('tenant_id', tid).maybeSingle()
  if (!wh) return c.json({ error: 'Webhook not found' }, 404)
  const payload = JSON.stringify({ event: 'test', data: { message: 'Test webhook from Nanowork' } })
  const sig = await hmacHex(wh.secret, payload)
  try {
    const resp = await fetch(wh.url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', 'X-Nanowork-Signature': sig },
      body: payload,
    })
    await sb.from('nano_webhooks').update({ last_fired_at: new Date().toISOString() }).eq('id', wh.id)
    return c.json({ delivered: resp.status < 400, status_code: resp.status })
  } catch (e) {
    return c.json({ delivered: false, error: String(e) })
  }
})

export default app
